storage: require min version file and deprecate cluster version key

RaduBerinde opened the pull request:
#### storage: require min version file

This change makes the min version file a requirement for opening an
existing store.

We have been writing this file out for at least two releases, so this
doesn't affect our compatibility story.

Release note: None
Epic: none

#### storage: deprecate cluster version key

This change deprecates the cluster version key. We have the version
outside the store (in the min version file) so we no longer need to
store it inside a KV.

To keep backward compatibility with 22.2, we still write the key but
we never read it. Once we stop supporting 22.2, we can stop writing it
altogether.

This required updating the `TestClusterVersionWriteSynthesize` to use
versions that work with the binary. We also clean up the test and move
it to `kvstorage` where it belongs.

Release note: None
Epic: none

Fixes #42653.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/97054)
<!-- Reviewable:end -->

tbg commented on pkg/storage/pebble.go:
```diff
@@ -1078,32 +1078,43 @@ func NewPebble(ctx context.Context, cfg PebbleConfig) (p *Pebble, err error) {
 
 	db, err := pebble.Open(cfg.StorageConfig.Dir, opts)
 	if err != nil {
+		if !minVerFileExists && strings.Contains(err.Error(), "already exists") {
```
`strings.Contains`? Wouldn't we want a structured error here? An argument could be made that it isn't worth it because it's such a dead code path & the event of a false positive isn't dramatic since the startup was failing either way?

RaduBerinde replied:
Yeah I would add a better way in Pebble. But for now I'm not sure of the higher level logic - what are the expectations if a machine fails during bootstrap? If it fails after creating the store but before creating the min version file, we would need manual cleanup before trying to bootstrap the store again. OTOH if we allow an existing store, then we might open an old version. Maybe the right thing is to add an ErrorIfNotEmpty mode to Pebble?

RaduBerinde replied:
Done.

tbg reviewed (approved):
Looks good to me (though I'd count this as a non-expert review).

jbowens reviewed (approved):
Nice

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 5 files at r1, 5 of 5 files at r2, 16 of 16 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/97054#-:-NOVKmWL5rFog_eUJci3:bmu9b66)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @renatolabs, @smg260, @sumeerbhola, and @tbg)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
TFTR!

RaduBerinde committed:
storage: deprecate cluster version key

This change deprecates the cluster version key. We have the version
outside the store (in the min version file) so we no longer need to
store it inside a KV.

To keep backward compatibility with 22.2, we still write the key but
we never read it. Once we stop supporting 22.2, we can stop writing it
altogether.

This required updating the `TestClusterVersionWriteSynthesize` to use
versions that work with the binary. We also clean up the test and move
it to `kvstorage` where it belongs.

Release note: None
Epic: none

RaduBerinde committed:
storage: require min version file

This change makes the min version file a requirement for opening an
existing store.

We have been writing this file out for at least two releases, so this
doesn't affect our compatibility story.

Release note: None
Epic: none

RaduBerinde commented:
bors r+
